Pull in latest from upstream with accepted hpa metrics change

ashleycutalo opened the pull request:
Grabbed the metrics we're interested in off of the models here https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/autoscaling/types.go

brancz committed:
*: cut v1.8.0-rc.0

(cherry picked from commit 89d2f9d66a6e2a1c285e63cc3f9ead3a77094aab)

tariq1890 committed:
pin go version to go mod artifact file

(cherry picked from commit ae2aecc4e65d09882ea7055b6ebb7ecf49c69539)

k8s-ci-robot committed:
Merge pull request #887 from tariq1890/release-1.8

(Cherry-pick) pin go version to go mod artifact file

ashleycutalo committed:
add hpa stats for current utilization and average value

ashleycutalo commented:
meow #compute @paulnivin

paulnivin commented:
let's include MinReplicas and MaxReplicas?

paulnivin commented on internal/store/hpa.go:
```diff
@@ -143,6 +143,51 @@ var (
+					Metrics: []*metric.Metric{
+						{
+							LabelValues: []string{strings.ToLower(string(c.Resource.Name))},
+							Value: float64(a.Status.CurrentMetrics.Resource.Current.Value),
```
CurrentMetrics is an array

ashleycutalo replied:
Fixed.

paulnivin commented on internal/store/hpa.go:
```diff
@@ -132,8 +132,8 @@ var (
-						metric.LabelKeys = []string{"condition", "status"}
-						metric.LabelValues = append(metric.LabelValues, string(c.Type))
+						metric.LabelKeys = []string{"condition", "status", "reason"}
+						metric.LabelValues = append(metric.LabelValues, string(c.Type), string(c.Reason))
```
raw data prior to this change:
```$ curl -s 10.45.31.74:8080/metrics | grep hpa_status_condition | grep dispatchavailability-main
kube_hpa_status_condition{namespace="dispatchavailability-production",hpa="dispatchavailability-main",condition="true",status="AbleToScale"} 1
kube_hpa_status_condition{namespace="dispatchavailability-production",hpa="dispatchavailability-main",condition="false",status="AbleToScale"} 0
kube_hpa_status_condition{namespace="dispatchavailability-production",hpa="dispatchavailability-main",condition="unknown",status="AbleToScale"} 0
kube_hpa_status_condition{namespace="dispatchavailability-production",hpa="dispatchavailability-main",condition="true",status="ScalingActive"} 1
kube_hpa_status_condition{namespace="dispatchavailability-production",hpa="dispatchavailability-main",condition="false",status="ScalingActive"} 0
kube_hpa_status_condition{namespace="dispatchavailability-production",hpa="dispatchavailability-main",condition="unknown",status="ScalingActive"} 0
kube_hpa_status_condition{namespace="dispatchavailability-production",hpa="dispatchavailability-main",condition="true",status="ScalingLimited"} 1
kube_hpa_status_condition{namespace="dispatchavailability-production",hpa="dispatchavailability-main",condition="false",status="ScalingLimited"} 0
kube_hpa_status_condition{namespace="dispatchavailability-production",hpa="dispatchavailability-main",condition="unknown",status="ScalingLimited"} 0
```

paulnivin commented on internal/store/hpa.go:
```diff
@@ -132,8 +132,8 @@ var (
 
```
addConditionMetrics(c.Status, c.Reason) is needed to pre-generate a matrix of all status w/ reason?

ashleycutalo replied:
Reason is an open-ended string unlike Status so I don't think we can generate a matrix off of it.

terinjokes committed:
deployment status conditions

Deployments, like Nodes, have status conditions observing the
current state. While the state of Available and Progressing conditions
can likely be inferred by other metrics, the state of ReplicaFailure can
not be inferred.

This changelist adds a new metric `kube_deployment_status_condition`
that observes all the conditions on a deployment for each condition
status. This is analogous to the status conditions observed by nodes and
horizontal pod autoscalers, and allows kube-state-metrics to observe
status conditions added by third-parties.

As an example, for a deployment that has stalled, the following new
metrics observed would allow an operator to detect the condition:

    kube_deployment_status_condition{deployment="example", namespace="default", condition="ReplicaFailure", status="true"} 1
    kube_deployment_status_condition{deployment="example", namespace="default", condition="ReplicaFailure", status="false"} 0
    kube_deployment_status_condition{deployment="example", namespace="default", condition="ReplicaFailure", status="unknown"} 0

Bug: #886
Signed-off-by: Terin Stock <terin@cloudflare.com>

terinjokes committed:
correct mapping of hpa condition status labels

The "condition" and "status" labels for the hpa status conditions were
mapped to the incorrect values. This resulted in the status being in the
condition label, and the condition in the status label.

This changelist corrects the mapping, so that condition and status map
to their respective values.

    kube_hpa_status_condition{condition="AbleToScale",hpa="hpa1",namespace="ns1",status="false"} 0
    kube_hpa_status_condition{condition="AbleToScale",hpa="hpa1",namespace="ns1",status="true"} 1
    kube_hpa_status_condition{condition="AbleToScale",hpa="hpa1",namespace="ns1",status="unknown"} 0

Fixes: f9658ca ("Add hpa conditions")
Signed-off-by: Terin Stock <terin@cloudflare.com>

terinjokes committed:
correct mapping of hpa condition status labels

The "condition" and "status" labels for the hpa status conditions were
mapped to the incorrect values. This resulted in the status being in the
condition label, and the condition in the status label.

This changelist corrects the mapping, so that condition and status map
to their respective values.

    kube_hpa_status_condition{condition="AbleToScale",hpa="hpa1",namespace="ns1",status="false"} 0
    kube_hpa_status_condition{condition="AbleToScale",hpa="hpa1",namespace="ns1",status="true"} 1
    kube_hpa_status_condition{condition="AbleToScale",hpa="hpa1",namespace="ns1",status="unknown"} 0

Fixes: f9658ca ("Add hpa conditions")
Signed-off-by: Terin Stock <terin@cloudflare.com>
(cherry picked from commit 6b09c5eb7d671c5faa7ef91a07ebc88bfb5ad9d1)

k8s-ci-robot committed:
Merge pull request #889 from terinjokes/patches/deployment-conditions

deployment status conditions

k8s-ci-robot committed:
Merge pull request #890 from terinjokes/patches/hpa-conditions

correct hpa condition status

milesbxf committed:
store: export NetworkPolicy objects

tariq1890 committed:
bump golangci-lint to latest stable release

tariq1890 committed:
fix shell script bug in licensecheck logic

YaoZengzeng committed:
framework for e2e test implemented in golang

Signed-off-by: YaoZengzeng <yaozengzeng@huawei.com>

k8s-ci-robot committed:
Merge pull request #894 from tariq1890/bump_golangci

bump golangci-lint to latest stable release

k8s-ci-robot committed:
Merge pull request #895 from tariq1890/fix_licensecheck

fix shell script bug in licensecheck logic

tariq1890 committed:
*: cut v1.8.0-rc.1

k8s-ci-robot committed:
Merge pull request #898 from tariq1890/release-1.8

*: cut v1.8.0-rc.1

YaoZengzeng committed:
fix hyperlink issue

Signed-off-by: YaoZengzeng <yaozengzeng@huawei.com>

k8s-ci-robot committed:
Merge pull request #900 from YaoZengzeng/url

fix hyperlink issue

k8s-ci-robot committed:
Merge pull request #896 from YaoZengzeng/framework

framework for e2e test implemented in golang

tariq1890 committed:
update k8s manifests to point to 1.8.0-rc.1

k8s-ci-robot committed:
Merge pull request #904 from tariq1890/rc_image

update k8s manifests to point to 1.8.0-rc.1

brancz committed:
*: Generate all manifests

Sharding and autosharding introduced a number of manifests that are
very similar, but not identical to the standard manifests. This
introduces generating all manifests using jsonnet and moves them to
the explicit `/examples` directory.

natalysheinin committed:
add mutatingwebhookconfiguration, validatingwebhookconfiguration metrics

YaoZengzeng committed:
[e2e] lint the metrics of ksm

Signed-off-by: YaoZengzeng <yaozengzeng@huawei.com>

k8s-ci-robot committed:
Merge pull request #907 from YaoZengzeng/lint

[e2e] lint the metrics of ksm

lilic committed:
VERSION: Update version to v1.8.0-rc.1

k8s-ci-robot committed:
Merge pull request #909 from LiliC/update-version

VERSION: Update version to v1.8.0-rc.1

tariq1890 committed:
*: cut v1.8.0

k8s-ci-robot committed:
Merge pull request #910 from tariq1890/cut1.8

*: cut v1.8.0

tariq1890 committed:
Merge branch 'release-1.8'

# Conflicts:
#	CHANGELOG.md
#	README.md
#	VERSION
#	go.mod

gtorre committed:
Update README

tariq1890 committed:
Merge pull request #911 from tariq1890/merge-1.8

Merge release-1.8 into master

k8s-ci-robot committed:
Merge pull request #912 from gtorre/patch-1

Update README

u5surf committed:
Add godoc badge to README.md #916

moideen committed:
Update golangci version to latest stable 1.19.1

k8s-ci-robot committed:
Merge pull request #926 from moideen/moideen-patch-1

Update golangci version to latest stable 1.19.1

k8s-ci-robot committed:
Merge pull request #921 from u5surf/issue-916

Add godoc badge to README.md #916

jukie committed:
Install tools as go modules

k8s-ci-robot committed:
Merge pull request #908 from brancz/generate

*: Generate all manifests

jiyeonseo committed:
fix broken link to './kubernetes/autosharding'

jiyeonseo committed:
Merge branch 'master' into master

k8s-ci-robot committed:
Merge pull request #928 from jiyeonseo/master

fix a broken link to './kubernetes/autosharding'

natalysheinin committed:
add instructions for adding a new k8s resource

clickyotomy committed:
Add defaults to ReclaimPolicy and VolumeBindingMode.

k8s-ci-robot committed:
Merge pull request #914 from natalysheinin/natalysheinin-add-webhookconfigurations

Add support for MutatingWebhookConfiguration +ValidatingWebhookConfiguration

manoelstilpen committed:
Update CONTRIBUTING.md

manoelstilpen committed:
Update CONTRIBUTING.md

k8s-ci-robot committed:
Merge pull request #929 from clickyotomy/master

Add check for ReclaimPolicy, VolumeBindingMode storageclass metrics.

bvwells committed:
Enable unconvert linter

manoelstilpen committed:
updates kube-state-metrics link; adds k8s channel link

k8s-ci-robot committed:
Merge pull request #934 from bvwells/unconvert

Enable unconvert linter

k8s-ci-robot committed:
Merge pull request #933 from manoelstilpen/master

adds slack channel link to CONTRIBUTING.md

bvwells committed:
Remove unnecessary type conversion

bvwells committed:
Clean up whitelist/blacklist

bvwells committed:
Simplify collector/metric set

k8s-ci-robot committed:
Merge pull request #935 from bvwells/unconvert-2

Remove unnecessary type conversion

k8s-ci-robot committed:
Merge pull request #936 from bvwells/lists

Pre-allocation of slices

simonpasquier committed:
Fix typo in the job metrics description

k8s-ci-robot committed:
Merge pull request #938 from simonpasquier/fix-job-metrics-typo

Fix typo in the job metrics description

k8s-ci-robot committed:
Merge pull request #927 from Jukie/tool-modules-918

Add tools as go modules

vjdhama committed:
update rolebinding api version in README

k8s-ci-robot committed:
Merge pull request #944 from vjdhama/master

update rolebinding api version in README

k8s-ci-robot committed:
Merge pull request #940 from natalysheinin/add-setup-documentation

[documentation] Adding a new Kubernetes Resource

JensErat committed:
add VolumeAttachment collector

Kubernetes has a new resource type: `VolumeAttachments`. They provide
helpful information on where a volume is attached and to alert on
unexpected attachment status (for example, differences between
information scraped from node-exporter and kube-state-metrics).

The collector adds a bunch of new metrics. Each VolumeAttachment (ie.,
each CSI-attached volume) will have one of each, so we do not overly
pollute the metrics space. Most metrics are rather unsurprising.

- `kube_volumeattachment_status_attachment_metadata`: provides a
  label-like export of the attachment metadata map. Generalizing the
  label-conversion function slightly helps at providing this metric.
- `kube_volumeattachment_created`: as VolumeAttachments are
  automatically created and we already suffered from duplicate
  `VolumeAttachments`, this can be invaluable for debugging
  misattachments.
- `kube_volumeattachment_spec_source_persistentvolume`: will only be
  generated when the volume source is of `PersistentVolume` type. The
  other type `inlineVolumeSpec` is still alpha-level and hard to map to
  metrics.

No end-to-end test manifest was added, as `VolumeAttachment`s are
automatically generated when mounting volumes.

Signed-off-by: Jens Erat <email@jenserat.de>

sinhaaks committed:
-m ADD kube_persistentvolumeclaim_status_condition metric for information about status of different conditions of persistent volume claim.

k8s-ci-robot committed:
Merge pull request #946 from c445/jenserat/volumeattachments

add VolumeAttachment collector

tariq1890 committed:
Merge branch 'master' into networkpolicy-collector

k8s-ci-robot committed:
Merge pull request #893 from monzo/networkpolicy-collector

store: collector for NetworkPolicy objects

tariq1890 committed:
clean up go mod indirect entries

tariq1890 committed:
move tools.go to the right directory

k8s-ci-robot committed:
Merge pull request #954 from tariq1890/gomod_clean

clean up go mod indirect entries

k8s-ci-robot committed:
Merge pull request #955 from tariq1890/mv_tools

move tools.go to the right directory

brancz committed:
README.md: Update recommendation on Prometheus discovery

k8s-ci-robot committed:
Merge pull request #959 from brancz/update-scrape-config

README.md: Update recommendation on Prometheus discovery

k8s-ci-robot committed:
Merge pull request #952 from ankitsinhaiiitg/aks-resizingStatusMetric

Add metric to show details of status condition of PVC resizing.

ashleycutalo commented:
meow #compute
